### PR TITLE
added metadata for several instructions

### PR DESCRIPTION
**Pull request by mcaldwelva**
Added some things needed to accept mcsema & dagger lifted IR

### CODE REVIEW

File: asm/internal/ll.bnf [hunk 1]
```diff
@@ -504,6 +504,7 @@ NamedType
 Value
 	: LocalIdent
 	| Constant
+	| "undef"     << nil, nil >>
```
Comment (mewmew) [inline asm/internal/ll.bnf]: Going forward we may need to introduce a notion for undefined values that is propagated through the AST and IR representations. For now, we can use `nil` simply to allow parsing of LLVM IR assembly files which contain `undef` values.

File: asm/internal/ll.bnf [hunk 2]
```diff
@@ -569,6 +570,7 @@ CharArrayConst
 
 StructConst
 	: "{" Elems "}"   << astx.NewStructConst($1) >>
+	| "<" "{" Elems "}" ">"  << astx.NewStructConst($2) >>
```
Comment (mewmew) [inline asm/internal/ll.bnf]: Thanks! had entirely forgotten about the packed structure literal.

File: asm/internal/ll.bnf [hunk 3]
```diff
@@ -1207,7 +1209,7 @@ GetElementPtrInst
 
 Indices
 	: empty
-	| "," IndexList   << $1, nil >>
+	| "," IndexList CommaAttachedMDs  << $1, nil >>
```
Comment (mewmew) [inline asm/internal/ll.bnf]: Indicies is not an instruction, and should not have attached metadata. The `CommaAttachedMDs` non-terminal belongs to the `GetElementPtrInst` production rule.

Or did you place it here to prevent a LR1 collision? If so annotate with a comment along the lines of:
```go
// TODO: Clean up when the parser generator no longer introduces ambiguities
// through the limitation of 1 rune lookahead.
//
// Structured in this way to allow for naiive 1 rune lookahead parser
// generators.
```

File: asm/internal/ll.bnf [hunk 4]
```diff
@@ -1369,7 +1371,12 @@ SelectInst
 // ~~~ [ call ] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
 CallInst
-	: "call" FastMathFlags ParamAttrs Type Value "(" Args ")" FuncAttrs   << astx.NewCallInst($3, $4, $6) >>
+	: OptTail "call" FastMathFlags OptCallConv ParamAttrs Type Value "(" Args ")" FuncAttrs CommaAttachedMDs   << astx.NewCallInst($5, $6, $8) >>
```
Comment (mewmew) [inline asm/internal/ll.bnf]: Thanks! The annotation for calling convention will come in use later by decomp! At that point we can propagate the data along to the IR layer through `astx.NewCallInst`.

File: asm/internal/ll.bnf [hunk 5]
```diff
@@ -1719,8 +1726,7 @@ OptAllocSizeNElems
 ;
 
 Elems
-	: empty
-	| ElemList
+	: ElemList
```
Comment (mewmew) [inline asm/internal/ll.bnf]: Why did you remove `empty`? The convention used to this point is that `Foos` represent 0 or more `Foo`s and `FooList` represents 1 or more `Foo`s.

File: asm/internal/ll.bnf [hunk 6]
```diff
@@ -1770,7 +1776,7 @@ CallConv
 	| "spir_kernel"
 	| "swiftcc"
 	| "webkit_jscc"
-	| "x86_64_sysvcc"
+	| "x86_64_sysvcc"      << ast.CallConvX86_64SysV, nil >>
```
Comment (mewmew) [inline asm/internal/ll.bnf]: The full list of calling conventions were added in rev f0c1093fae039472e8f4186a80850252d02583e2. Sorry, I had forgotten to push this to GitHub.

**Comment by mewmew** [comment]
Hi @mcaldwelva!

Thanks a lot for adding the attached metadata to instructions. It's been on my mental todo the last couple of weeks. I think it will allow us to propagate a lot of useful information through the decompilation pipeline in the future.

As mentioned in the review comment above, I had forgotten to push the full list of calling conventions that were added in rev f0c1093fae039472e8f4186a80850252d02583e2. Would you mind syncing with master so that the merge conflicts gets resolved?

Cheers /u

**Comment by mewmew** [comment]
Oh, and one more thing. The [llir project is released into the public domain](https://github.com/llir/llvm#public-domain). To simplify usage for anyone in the future, would you be alright with releasing your contribution into the public domain as well?

File: asm/internal/ll.bnf [hunk 7]
```diff
@@ -1610,7 +1617,7 @@ UnnamedAddr
 
 OptCommaAlign
 	: empty
-	| "," Align
+	| "," Align CommaAttachedMDs
```
Comment (mewmew) [inline asm/internal/ll.bnf]: Same as with https://github.com/llir/llvm/pull/21/files#r115116240

`OptCommaAlign` is not an instruction, and if possible the attached metadata should be added to the production rule of the instruction, specifically for optional alignment, the `load`, `store` and `alloca` instructions.

If you wish, we could merge it as is, and I'll look into moving the attached metadata to the relevant instructions?

File: asm/internal/ll.bnf (hunk 5 above)
Comment (mcaldwelva) [inline asm/internal/ll.bnf]: After I modified StructConst, this caused an LR1 collision. I tried walking through the productions and it seemed consistent, so I'm not sure how to put this back in.

File: asm/internal/ll.bnf (hunk 3 above)
Comment (mcaldwelva) [inline asm/internal/ll.bnf]: Yep, I put it here to avoid collision.

File: asm/internal/ll.bnf (hunk 7 above)
Comment (mcaldwelva) [inline asm/internal/ll.bnf]: Sorry, this was a bit of a kludge, but again trying to work around LR1 limitations. I'll add the comments for now at least.

File: asm/internal/ll.bnf (hunk 5 above)
Comment (mewmew) [inline asm/internal/ll.bnf]: > After I modified StructConst, this caused an LR1 collision. I tried walking through the productions and it seemed consistent, so I'm not sure how to put this back in.

Thanks for the background, I'll take a look and see what the cause may be.

**Comment by mewmew** [comment]
Hi @mcaldwelva!

I think we are getting ready to merge. Just for public record, would you be willing to release your contributions to llir to the public domain?

I know you've already done so for the decomp project in https://github.com/decomp/decomp/pull/189#issuecomment-299670801

Cheers /u

**Comment by mcaldwelva** [comment]
Happy to give this to the public domain as well!

**Comment by mewmew** [comment]
Lovely! Merged, and done : )
